### DISCUSSION TITLE
Georeferenced predictions

### DISCUSSION THREAD

**Discussion by lcoandrade** [torchgeo/torchgeo#1745] [Q&A]
Is there any advances towards a transparent way to make georeferenced predictions? So far, I'm working with Rasterio to accomplish that. The problem is that my process is taking way longer than making the georeferenced prediction using RasterVision, for instance.

**Comment by adamjstewart** [comment]
Not yet, but this is something I would really like to see if anyone wants to try to take a stab at it. #1407 and #1440 contain a list of hurdles and suggestions for this. If we can keep the CRS/transform and other metadata in the R-tree and pass it through Lightning, we should be able to implement a `predict_step` that saves the output to a GeoTIFF.

**Comment by lcoandrade** [comment]
I haven't taken a look at how RasterVision does that. Do you know what was their approach?

**Reply by adamjstewart** [reply]
Not sure how they do it, maybe @lewfish @AdeelH can comment

**Reply by AdeelH** [reply]
For semantic segmentation, Raster Vision aggregates the predictions for each window in a numpy array and then writes that array out to a GeoTIFF with the same extent, projection etc. as the input GeoTiff. This functionality is spread across [`SemanticSegmentationLabels`](https://github.com/azavea/raster-vision/blob/master/rastervision_core/rastervision/core/data/label/semantic_segmentation_labels.py) and [`SemanticSegmentationLabelStore`](https://github.com/azavea/raster-vision/blob/master/rastervision_core/rastervision/core/data/label_store/semantic_segmentation_label_store.py).

It should theoretically be possible to use TorchGeo-produced predictions with this functionality by appropriately modifying [this example](https://docs.rastervision.io/en/stable/usage/tutorials/lightning_workflow.html#Make-predictions-for-scene). You would need an iterable of [Raster Vision windows](https://docs.rastervision.io/en/stable/api_reference/_generated/rastervision.core.box.Box.html), an iterable of corresponding predictions, a [`ClassConfig`](https://docs.rastervision.io/en/stable/api_reference/_generated/rastervision.core.data.class_config.ClassConfig.html#classconfig), and a [`CRSTransformer`](https://docs.rastervision.io/en/stable/api_reference/_generated/rastervision.core.data.crs_transformer.rasterio_crs_transformer.RasterioCRSTransformer.html#rastervision.core.data.crs_transformer.rasterio_crs_transformer.RasterioCRSTransformer.from_uri).

P.S. @adamjstewart, FYI Lewis is no longer involved in the project.

**Reply by lcoandrade** [reply]
The idea I'm using so far is this (using Rasterio `from_origin` and `merge`):

1. I create a Grid Dataloader (RasterDataset+GridGeoSampler+Dataloader) (something parallel to SlidingWindowGeoDataset);
2. For each sample, I create a geotiff with the CRS (from the dataset) and geotransform matrix using `from_origin(bbox.minx, bbox.maxy, pixel_size, pixel_size)`;
3. I add each chipped geotiff to a list;
4. I merge the list of geotiffs to make the full geotiff prediction;
5. Save this as a file.

Do you see a bottleneck in this approach?
